File: Sudoku_cpp/src/Sudoku/Sudoku.cpp
#include <Sudoku.hpp>

#include <unistd.h>

Sudoku::Sudoku() : Grid() {
  this->min_num = 1;
  this->max_num = 9;
  this->space = 0;
  this->revisable_grid_ = new bool[GRID_SIZE * GRID_SIZE];
}

Sudoku::Sudoku(size_t grid_size, size_t box_size, int min_num, int max_num, int space) : Grid(grid_size, box_size){
  this->min_num = min_num;
  this->max_num = max_num;
  this->space = space;
  this->revisable_grid_ = new bool[GRID_SIZE * GRID_SIZE];
}

Sudoku::Sudoku(size_t grid_size, size_t box_size, std::string grid_str, int min_num, int max_num, int space) : Grid(grid_size, box_size, grid_str) {
  // int* g = parse_str(grid_str);
  // for(int i = 0; i < grid_size * grid_size; i++) {
  //   std::cout << "int: " << g[i] << '\n';
  // }
  // std::cout << "g: " << &g << '\n';
  this->min_num = min_num;
  this->max_num = max_num;
  this->space = space;
  this->revisable_grid_ = new bool[grid_size * grid_size];
  for(int i = 0; i < grid_size * grid_size; i++) {
    if(this->grid_[i] == this->space)
      this->revisable_grid_[i] = true;
    else
      this->revisable_grid_[i] = false;
  }
}

// copy construct (deep copy)
Sudoku::Sudoku(const Sudoku& s) {
  this->GRID_SIZE = s.GRID_SIZE;
  this->BOX_SIZE = s.BOX_SIZE;
  this->min_num = s.min_num;
  this->max_num = s.max_num;
  this->space = s.space;

  delete this->grid_;
  this->grid_ = new int[GRID_SIZE * GRID_SIZE];
  for(int i = 0; i < GRID_SIZE * GRID_SIZE; i++)
    this->grid_[i] = s.grid_[i];
    
  this->revisable_grid_ = new bool[GRID_SIZE * GRID_SIZE];
  for(int i = 0; i < GRID_SIZE * GRID_SIZE; i++) {
    if(this->grid_[i] == this->space)
      this->revisable_grid_[i] = true;
    else
      this->revisable_grid_[i] = false;
  }
}

bool Sudoku::write(size_t row, size_t col, int num) {
    if (this->revisable_grid_[row * GRID_SIZE + col] &&
        (row >= 0 && row < GRID_SIZE) &&
        (col >= 0 && col < GRID_SIZE) &&
        ((num >= min_num && num <= max_num) || num == space)) {
        this->grid_[row * GRID_SIZE + col] = num;
        return true;
    }
    return false;
}

void Sudoku::reset(size_t grid_size, size_t box_size, std::string grid_str, int min_num, int max_num, int space) {
  this->GRID_SIZE = grid_size;
  this->BOX_SIZE = box_size;

  if (this->grid_ != nullptr)
    delete this->grid_;
  this->grid_ = parse_str(grid_str);


  this->min_num = min_num;
  this->max_num = max_num;
  this->space = space;

  for(int i = 0; i < GRID_SIZE * GRID_SIZE; i++) {
    if(this->grid_[i] == this->space)
      this->revisable_grid_[i] = true;
    else
      this->revisable_grid_[i] = false;
  }
}

void Sudoku::show() {
  std::cout << "Sudoku Information:" << std::endl;
  std::cout << "GRID_SIZE: " << GRID_SIZE << std::endl;
  std::cout << "BOX_SIZE: " << BOX_SIZE << std::endl;
  std::cout << "min_num: " << min_num << std::endl;
  std::cout << "max_num: " << max_num << std::endl;
  std::cout << "space: " << space << std::endl;
  
  // std::cout << "show grid_: " << &(this->grid_) << '\n';
  this->Grid::show();
  // // std::cout << "[";
  // for(int i = 0; i < GRID_SIZE; i++){
  //   std::cout << "[";
  //   for(int j = 0; j < GRID_SIZE; j++){
  //     std::cout << this->grid_[i * GRID_SIZE + j] << ", ";
  //   }
  //   std::cout << "],";
  //   // if(i != GRID_SIZE - 1)
  //     std::cout << "\n";
  // }
  // // std::cout << "]\n";
}

bool Sudoku::check_repeat(int* array) {
  int num = GRID_SIZE + 1;
  int* repeat = new int[num]();
  bool res = true;
  for(int i = 0; i < sizeof(array) / sizeof(array[0]); i++)
    repeat[array[i] - min_num + 1] += 1;
  for(int i = 1; i < num; i++) {
    if(repeat[i] > 1)
      res = false;
  }
  delete repeat;
  return res;
}

bool Sudoku::valid() {
  bool flag = true;
  for(int i = 0; i < GRID_SIZE; i++)
    flag = flag && check_repeat(get_row(i));
  
  for(int i = 0; i < GRID_SIZE; i++)
    flag = flag && check_repeat(get_column(i));

  for(int i = 0; i < GRID_SIZE / BOX_SIZE; i++) {
    for(int j = 0; j < GRID_SIZE / BOX_SIZE; j++)
      flag = flag && check_repeat(get_box(i * BOX_SIZE, j * BOX_SIZE));
  }
  return flag;
}


std::pair<int, int> Sudoku::get_first_space() {
  std::pair<int, int> res;
  for(int i = 0; i < GRID_SIZE; i++) {
    for(int j = 0; j < GRID_SIZE; j++) {
      if(grid_[i * GRID_SIZE + j] == space) {
        res.first = i;
        res.second = j;
        return res;
      }
    }
  }
  res = std::make_pair<int, int>(-1, -1);
  return res;
}

std::pair<int, int> Sudoku::get_next_space(size_t r, size_t c) {
  std::pair<int, int> res;
  for(int j = c + 1; j < GRID_SIZE; j++) {
    if(grid_[r * GRID_SIZE + j] == space) {
      res.first = r;
      res.second = j;
      return res;
    }
  }

  for(int i = r + 1; i < GRID_SIZE; i++) {
    for(int j = 0; j < GRID_SIZE; j++) {
      if(grid_[i * GRID_SIZE + j] == space) {
        res.first = i;
        res.second = j;
        return res;
      }
    }
  }
  res = std::make_pair<int, int>(-1, -1);
  return res;
}

std::set<size_t> Sudoku::get_candidate(size_t r, size_t c) {
  std::set<size_t> res;
  
  int* array = new int[GRID_SIZE]();
  int* row = get_row(r);
  int* col = get_column(c);
  for(int i = 0; i < GRID_SIZE; i++) {
    if (row[i] != this->space)
      array[row[i] - min_num] = 1;
    if (col[i] != this->space)
      array[col[i] - min_num] = 1;
  }

  int* box = get_box(r, c);
  for(int i = 0; i < BOX_SIZE * BOX_SIZE; i++){
    if (box[i] != this->space) 
      array[box[i] - min_num] = 1;
 
  }
  for(int i = 0; i < GRID_SIZE; i++) {
    if(array[i] == 0)
      res.insert(i + min_num);
  }

  delete row;
  delete col;
  delete box;
  delete array;
  return res;
}


void Sudoku::inference() {
  std::pair<int, int> s = get_first_space();
  std::cout << "inference begin" << std::endl;
  while (s.first != -1 && s.second != -1) {
    std::set<size_t> candidate = get_candidate(s.first, s.second);
    if (candidate.size() != 0) {
      std::cout << "candidates of (" << s.first << ", " << s.second << "): ";
      std::cout << '{';
      for(auto& it : candidate)
        std::cout << it << ", ";
      std::cout << '}' << std::endl;
    }
    s = get_next_space(s.first, s.second);
  }
  std::cout << "inference end" << std::endl;
}

bool Sudoku::fill(size_t r, size_t c) {
  std::set<size_t> candidate = get_candidate(r, c);
  std::pair<int, int> next;
  for(auto it : candidate) {
    grid_[r * GRID_SIZE + c] = it;
    // sleep(1);
    // std::cout << "(" << r << ", " << c<< ") ";
    // std::cout << "candidate: " << it << '\n';
    // this->Grid::show();
    next = get_next_space(r, c);
    if (next.first == -1 && next.second == -1)
      return true;
    else {
      if (fill(next.first, next.second))
        return true;
      else
        grid_[r * GRID_SIZE + c] = space;
    }
  }
  return false;
}

bool Sudoku::solve() {
  if(!valid())
    return false;
  std::pair<int, int> rc = get_first_space();
  bool res = fill(rc.first, rc.second);
  // std::cout << "res: " << res << '\n';
  // std::cout << "valid: " << valid() << '\n';
  if(res && valid()){
    std::cout << "Solve!\n";
    return true;
  }
  else {
    std::cout << "Fail\n";
    return false;
  }
}

File: Sudoku_py/Sudoku.py
import pickle
import numpy as np
import copy

from Grid import Grid

class Sudoku(Grid):
  '''
  function:
    construct and initialize
  '''
  def __init__(self, grid_size, box_size, sudoku_str, min_num, max_num, space):
    grid = self.parse_str(grid_size, sudoku_str)

    super(Sudoku, self).__init__(grid_size, box_size, grid)
    self.min_num = min_num
    self.max_num = max_num
    self.space = space
    self.revisable_grid = (self.grid == self.space)

  '''
  function:
    parse str sudoku
  Output:
    np.array(dtype = int, shape = (grid_size, grid_size))
  '''
  def parse_str(self, grid_size, sudoku_str):
    sudoku_list = []
    for i in sudoku_str:
      sudoku_list.append(int(i))
    return np.array(sudoku_list).reshape(grid_size, grid_size)
  
  '''
  function:
    transform sudoku into str
  Output:
    str
  '''
  def __str__(self):
    sudoku_str = ''
    for i in self.grid.flatten():
      sudoku_str += str(i)
    return sudoku_str

  '''
  function:
    return a deep copy instance
  Output:
    sudokuku
  '''
  def clone(self):
    return copy.deepcopy(self)
  
  '''
  function:
    return object in the format of serial data
  Output:
    bytes
  '''
  def serialize(self):
    return pickle.dumps(self)

  '''
  function:
    write number to given cell if it's valid
  '''
  def write(self, r, c, num):
    if (self.revisable_grid[r][c] and
        (r >=0 and r < self.GRID_SIZE) and
        (c >=0 and c < self.GRID_SIZE) and
        ((num <= self.max_num and num >= self.min_num) or num == self.space)):
      self.grid[r][c] = num
      return True
    else:
      return False

  '''
  function:
    reset rule and data of sudoku
  '''
  def reset_sudoku(self, grid_size, box_size, sudoku_str, min_num, max_num, space):
    self.GRID_SIZE = grid_size
    self.BOX_SIZE = box_size
    self.grid = self.parse_str(sudoku_str)
    self.min_num = min_num
    self.max_num = max_num
    self.space = space
    self.revisable_grid = (self.grid == self.space)

  '''
  function:
    show sudoku information
  '''
  def show(self):
    print("Sudoku Information:")
    print("GRID_SIZE: {}".format(self.GRID_SIZE))
    print("BOX_SIZE: {}".format(self.BOX_SIZE))
    print("min_num: {}".format(self.min_num))
    print("max_num: {}".format(self.max_num))
    print("space: {}".format(self.space))
    print(self.grid)
  
  '''
  function:
    check whether the array contains repeated nonzero values
  Input:
    np.array
  Output:
    bool
  '''
  def check_repeat(self, array):
    unique_value, repeat = np.unique(array, return_counts = True)
    for i in range(len(repeat)):
      if (repeat[i] != 1 and 
          unique_value[i] != self.space):
        return False
    return True

  '''
  function:
    check the validity of the sudoku
  Output:
    bool: valid or not
  '''
  def valid(self):
    flag = True
    # for all row
    for r in range(self.GRID_SIZE):
      row = self.get_row(r)
      flag = flag and self.check_repeat(row)

    # for all column
    for c in range(self.GRID_SIZE):
      column = self.get_column(c)
      flag = flag and self.check_repeat(column)

    # for all block
    for r in range(self.GRID_SIZE // self.BOX_SIZE):
      for c in range(self.GRID_SIZE // self.BOX_SIZE):
        box = self.get_box(r * self.BOX_SIZE, c * self.BOX_SIZE).flatten()
        flag = flag and self.check_repeat(box)

    return flag

  '''
  function:
    return the first space of the Sudoku
  Output:
    (row, column)
    (-1, -1) means can't find the first space, sudoku has been solved
  '''
  def get_first_space(self):
    for i in range(self.GRID_SIZE):
      for j in range(self.GRID_SIZE):
        if self.grid[i][j] == self.space:
          return i, j
        
    # can't find space 
    return -1, -1

  '''
  function:
    return the next space from given current space
  Input:
    row: row of current space
    column: column of current space
  Output:
    (next_row, next_column)
    (-1, -1) means can't find next space, sudoku has been solved
  '''
  def get_next_space(self, r, c):
    # space in the same row
    for y in range(c + 1, self.GRID_SIZE):
      if self.grid[r][y] == self.space:
        return r, y

    # space in other row
    for x in range(r + 1, self.GRID_SIZE):
      for y in range(self.GRID_SIZE):
        if self.grid[x][y] == self.space:
          return x, y
    
    # can't find space 
    return -1, -1

  '''
  function:
    return all candidates of given space
  Input:
    row: row of space
    column: column of space
  Output:
    set{candidate1, candidate2, ...}
  '''
  def get_candidate(self, r, c):
    res = set([x for x in range(self.min_num, self.max_num + 1)])
    res -= set(self.get_row(r))
    res -= set(self.get_column(c))
    res -= set(self.get_box(r, c).flatten())
    return res

  '''
  function:
    return the spaces with only one candidate and their candidate value
  Output:
    dict{(space_row, space_column): candidate}
  '''
  def get_space_with_one_candidate(self):
    candidate_dict = {}

    space = self.get_first_space()
    while (space != (-1, -1)):
      candidate = self.get_candidate(space[0], space[1])
      if (len(candidate) == 1):
        for v in candidate:  
          candidate_dict[space] = v
      space = self.get_next_space(space[0], space[1])
    return candidate_dict
  
  '''
  function:
    infer and show the candidates for all spaces
  '''
  def inference(self):
    s = self.get_first_space()
    print("inference begin")
    while (s != (-1, -1)):
      candidate = self.get_candidate(s[0], s[1])
      if (len(candidate) != 0):
        print("candidates of {}: {}".format(s, candidate))
      s = self.get_next_space(s[0], s[1])
    print("inference end")

  '''
  function:
    fill space with only one candidate iterately,
    it can solve some easy sudoku and accelerate for difficult sudoku
  '''
  def ez_fill(self):
    candidate_dict = self.get_space_with_one_candidate()
    while(candidate_dict != {}):
      for k in candidate_dict.keys():
        self.grid[k[0]][k[1]] = candidate_dict[k]
      # print(self.grid)
      candidate_dict = self.get_space_with_one_candidate()

  '''
  function:
    fill space according to the candidates recursively,
    it can solve all sudoku
  '''
  def fill(self, r, c):
    for candidate in self.get_candidate(r, c):
      self.grid[r][c] = candidate
      next_r, next_c = self.get_next_space(r, c)
      if (next_r, next_c) == (-1, -1):
        return True
      elif self.fill(next_r, next_c):
        return True
      else:
        self.grid[r][c] = self.space
    return False
  
  '''
  function:
    solve sudoku
  Output:
    bool: solve or not
  '''
  def solve(self):
    if (not self.valid()):
      return False
    self.ez_fill()
    if (not self.valid()):
      return False
    r, c = self.get_first_space()
    if (self.fill(r, c) and self.valid()):
      print("Solve!")
      return True
    else: 
      print("Fail!")
      return False
